torchtitan/checkpoint.py

```diff
@@ -160,19 +211,124 @@ def save(self, curr_step: int, force: bool = False) -> None:
+                    )
+                    return False
+                elif curr_step % 5 == 4:
+                    self.time_sync_work.wait()
+                    self.time_sync_work = None
+                    time_sync_result = self.time_sync_result.item()
+                    self.time_sync_result = None
+                    if time_sync_result == 0:
+                        return False
+                else:
+                    return False
+
+        if self.time_sync_work:
+            self.time_sync_work.wait()
+            self.time_sync_work = None
+            self.time_sync_result = None
+
+        return True
+
+    def _async_wait(self) -> None:
+        if self.async_mode == AsyncMode.ASYNC_WITH_PINNED_MEM:
+            logger.debug(
+                f"Waiting for the background process to finish, {time.monotonic()=}.:.2f"
+            )
+            if not self.mp.is_alive():
+                raise RuntimeError("The checkpoint background process is dead.")
+            _ = self.mp_queue_recv.get()
+        elif self.async_mode == AsyncMode.ASYNC:
+            if self.async_future is not None:
+                self.async_future.result()
+
+    def _async_with_pinned_memory(self, checkpoint_id: str) -> None:
+        state_dict = dcp.state_dict_saver._stateful_to_state_dict(self.states)
+        if self.cpu_offload_state_dict is None:
+            logger.debug(f"Preparing the CPU memory, {time.monotonic()=}.:.2f")
+            self.cpu_offload_state_dict = _create_cpu_state_dict(
+                state_dict, pin_memory=True
+            )
+
+        logger.debug(f"Staging the state_dict, {time.monotonic()=}.:.2f")
+        with torch.cuda.stream(self.staging_stream):
+            self.cpu_offload_state_dict = _copy_state_dict(
+                state_dict,
+                self.cpu_offload_state_dict,
+                non_blocking=True,
+            )
+            self.staging = True
+            self.staging_state_dict = state_dict
+            self.staging_id = checkpoint_id
+
+    def save(self, curr_step: int, force: bool = False) -> None:
+        """
+        force = True will force the checkpoint to be saved, even if the interval
+        has not been reached.
+        This only happens when train_state.step == job_config.training.steps, or
+        for initial seed checkpoint.
+        """
+        if not self._should_save(curr_step, force):
+            return
+
+        begin = time.monotonic()
+        checkpoint_id = self._create_checkpoint_id(curr_step)
+        self._async_wait()
+        if force:
+            self._save_last_step(curr_step)
+        elif self.async_mode == AsyncMode.ASYNC_WITH_PINNED_MEM:
+            self._async_with_pinned_memory(checkpoint_id)
+        elif self.async_mode == AsyncMode.ASYNC:
+            self.async_future = dcp.async_save(
+                self.states, checkpoint_id=checkpoint_id, process_group=self.pg
```

Comment:
@fegin Why did you choose to use the GLOO process group for the async save? Is it expected to make this more efficient?

Neither the DCP docs nor https://discuss.pytorch.org/t/distributed-w-torchtitan-optimizing-checkpointing-efficiency-with-pytorch-dcp/211250 mention or recommend this.

I'm curious to know if this was on purpose and if you have any numbers to show

Thanks!